pkg/services/govmomi/service.go

```diff
@@ -441,6 +425,61 @@ func (vms *VMService) reconcileIPAddresses(ctx *virtualMachineContext) (bool, er
 	return true, nil
 }
 
+// resolveIpAddressNamesForClaims checks that all the IPAddressClaims have been
+// satisfied. If waiting on a claim to have an IPAddress, the status will refelct
+// how many are bound out of the total expected. A slice of slices matching the
+// structure of the device and AddressFromPools slices is returned so that
+// names don't need to be resolved more than once.
+func resolveIpAddressNamesForClaims(ctx *virtualMachineContext) ([][]string, error) {
+	boundClaims := 0
+	totalClaims := 0
+	ipAddrNames := make([][]string, len(ctx.VSphereVM.Spec.Network.Devices))
+	for devIdx, device := range ctx.VSphereVM.Spec.Network.Devices {
+		ipAddrNames[devIdx] = make([]string, len(device.AddressesFromPools))
+		for poolRefIdx := range device.AddressesFromPools {
+			// check if claim exists
+			ipAddrClaim := &ipamv1.IPAddressClaim{}
+			ipAddrClaimName := IPAddressClaimName(ctx.VSphereVM.Name, devIdx, poolRefIdx)
+			ipAddrClaimKey := apitypes.NamespacedName{
+				Namespace: ctx.VSphereVM.Namespace,
+				Name:      ipAddrClaimName,
+			}
+			var err error
+			ctx.Logger.V(5).Info("fetching IPAddressClaim", "name", ipAddrClaimKey.String())
+			if err = ctx.Client.Get(ctx, ipAddrClaimKey, ipAddrClaim); err != nil && !apierrors.IsNotFound(err) {
+				ctx.Logger.Error(err, "error fetching IPAddressClaim", "name", ipAddrClaimName)
+				return nil, err
```

Comment:
even with the error we should check for the state of the other IPAddressClaim objects and aggregate any errors at the end of the for loop, construct an error and return.

Reply:
While I agree with the spirit of this suggestion, I'm not sure I agree with this one specifically.

The prior step in this reconciliation loop does a find or create for the claims. 

In the current step of the loop, the controller is waiting for the claim to have an address bound. The case where the claim is not found is handled. The only reason this if condition would be hit is if there is an api error. Continuing on and attempting to look up other claims will likely fail too. Gathering up more api errors seemingly provides diminished value.

Unless you think it is important, we'll skip on this suggestion, and instead look for other places where we can apply the suggested strategy. 

WDYT?

Reply:
@srm09